File: core/schema-validate.js
/**
 * Lightweight JSON Schema validator (zero dependencies).
 * Supports: type, required, enum, const, minimum, maximum, pattern, oneOf, $ref (local $defs).
 */

const fs = require('fs');
const path = require('path');

function typeOf(value) {
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'array';
  return typeof value;
}

function matchesType(value, expected) {
  const types = Array.isArray(expected) ? expected : [expected];
  return types.some((t) => {
    if (t === 'integer') return Number.isInteger(value);
    if (t === 'number') return typeof value === 'number' && !Number.isNaN(value);
    return typeOf(value) === t;
  });
}

function resolveRef(ref, rootSchema) {
  if (!ref.startsWith('#/')) return null;
  const parts = ref.slice(2).split('/');
  let node = rootSchema;
  for (const part of parts) {
    node = node?.[part];
    if (!node) return null;
  }
  return node;
}

function validateNode(value, schema, rootSchema, pointer, errors) {
  if (!schema || typeof schema !== 'object') return;

  if (schema.$ref) {
    const resolved = resolveRef(schema.$ref, rootSchema);
    if (resolved) validateNode(value, resolved, rootSchema, pointer, errors);
    else errors.push(`${pointer}: unresolved $ref ${schema.$ref}`);
    return;
  }

  if (schema.oneOf) {
    const branchErrors = [];
    let validBranches = 0;
    for (const branch of schema.oneOf) {
      const branchErrs = [];
      validateNode(value, branch, rootSchema, pointer, branchErrs);
      if (branchErrs.length === 0) validBranches += 1;
      else branchErrors.push(branchErrs);
    }
    if (validBranches !== 1) {
      errors.push(`${pointer}: must match exactly one schema branch (matched ${validBranches})`);
    }
    return;
  }

  if (schema.type && !matchesType(value, schema.type)) {
    errors.push(`${pointer}: expected type ${JSON.stringify(schema.type)}, got ${typeOf(value)}`);
    return;
  }

  if (schema.enum && !schema.enum.includes(value)) {
    errors.push(`${pointer}: must be one of ${schema.enum.join(', ')}`);
  }

  if (schema.const !== undefined && value !== schema.const) {
    errors.push(`${pointer}: must equal ${JSON.stringify(schema.const)}`);
  }

  if (typeof value === 'number') {
    if (schema.minimum !== undefined && value < schema.minimum) {
      errors.push(`${pointer}: must be >= ${schema.minimum}`);
    }
    if (schema.maximum !== undefined && value > schema.maximum) {
      errors.push(`${pointer}: must be <= ${schema.maximum}`);
    }
  }

  if (typeof value === 'string' && schema.pattern) {
    const re = new RegExp(schema.pattern);
    if (!re.test(value)) errors.push(`${pointer}: must match pattern ${schema.pattern}`);
  }

  if (schema.required && typeOf(value) === 'object') {
    for (const key of schema.required) {
      if (value[key] === undefined) errors.push(`${pointer}: missing required property "${key}"`);
    }
  }

  if (schema.properties && typeOf(value) === 'object') {
    for (const [key, propSchema] of Object.entries(schema.properties)) {
      if (value[key] !== undefined) {
        validateNode(value[key], propSchema, rootSchema, `${pointer}/${key}`, errors);
      }
    }
  }

  if (schema.items && typeOf(value) === 'array') {
    value.forEach((item, i) => {
      validateNode(item, schema.items, rootSchema, `${pointer}[${i}]`, errors);
    });
  }
}

function validateAgainstSchema(data, schema) {
  const errors = [];
  validateNode(data, schema, schema, '', errors);
  return { valid: errors.length === 0, errors };
}

function loadSchema(schemaName) {
  const file = schemaName.endsWith('.json') ? schemaName : `${schemaName}.schema.json`;
  const schemaPath = path.join(__dirname, 'schemas', file);
  if (!fs.existsSync(schemaPath)) {
    throw new Error(`Schema not found: ${schemaPath}`);
  }
  return JSON.parse(fs.readFileSync(schemaPath, 'utf8'));
}

function validateWithNamedSchema(data, schemaName) {
  const schema = loadSchema(schemaName);
  return validateAgainstSchema(data, schema);
}

module.exports = {
  validateAgainstSchema,
  validateWithNamedSchema,
  loadSchema,
};

File: core/schema_validate.py
"""Lightweight JSON Schema validator (zero dependencies beyond stdlib)."""

import json
import os
import re


def _type_of(value):
    if value is None:
        return "null"
    if isinstance(value, list):
        return "array"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int) and not isinstance(value, bool):
        return "integer"
    if isinstance(value, (int, float)):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, dict):
        return "object"
    return type(value).__name__


def _matches_type(value, expected):
    types = expected if isinstance(expected, list) else [expected]
    actual = _type_of(value)
    for t in types:
        if t == "integer" and isinstance(value, int) and not isinstance(value, bool):
            return True
        if t == "number" and isinstance(value, (int, float)) and not isinstance(value, bool):
            return True
        if actual == t:
            return True
    return False


def _resolve_ref(ref, root_schema):
    if not ref.startswith("#/"):
        return None
    node = root_schema
    for part in ref[2:].split("/"):
        node = node.get(part) if isinstance(node, dict) else None
        if node is None:
            return None
    return node


def _validate_node(value, schema, root_schema, pointer, errors):
    if not isinstance(schema, dict):
        return

    if "$ref" in schema:
        resolved = _resolve_ref(schema["$ref"], root_schema)
        if resolved:
            _validate_node(value, resolved, root_schema, pointer, errors)
        else:
            errors.append(f"{pointer}: unresolved $ref {schema['$ref']}")
        return

    if "oneOf" in schema:
        valid = 0
        for branch in schema["oneOf"]:
            branch_errs = []
            _validate_node(value, branch, root_schema, pointer, branch_errs)
            if not branch_errs:
                valid += 1
        if valid != 1:
            errors.append(f"{pointer}: must match exactly one schema branch (matched {valid})")
        return

    if "type" in schema and not _matches_type(value, schema["type"]):
        errors.append(f"{pointer}: expected type {schema['type']}, got {_type_of(value)}")
        return

    if "enum" in schema and value not in schema["enum"]:
        errors.append(f"{pointer}: must be one of {', '.join(map(str, schema['enum']))}")

    if "const" in schema and value != schema["const"]:
        errors.append(f"{pointer}: must equal {json.dumps(schema['const'])}")

    if isinstance(value, (int, float)) and not isinstance(value, bool):
        if "minimum" in schema and value < schema["minimum"]:
            errors.append(f"{pointer}: must be >= {schema['minimum']}")
        if "maximum" in schema and value > schema["maximum"]:
            errors.append(f"{pointer}: must be <= {schema['maximum']}")

    if isinstance(value, str) and "pattern" in schema:
        if not re.match(schema["pattern"], value):
            errors.append(f"{pointer}: must match pattern {schema['pattern']}")

    if isinstance(value, dict):
        for key in schema.get("required", []):
            if key not in value:
                errors.append(f'{pointer}: missing required property "{key}"')
        for key, prop_schema in schema.get("properties", {}).items():
            if key in value:
                _validate_node(value[key], prop_schema, root_schema, f"{pointer}/{key}", errors)

    if isinstance(value, list) and "items" in schema:
        for i, item in enumerate(value):
            _validate_node(item, schema["items"], root_schema, f"{pointer}[{i}]", errors)


def validate_against_schema(data, schema):
    errors = []
    _validate_node(data, schema, schema, "", errors)
    return {"valid": len(errors) == 0, "errors": errors}


def load_schema(schema_name):
    file_name = schema_name if schema_name.endswith(".json") else f"{schema_name}.schema.json"
    schema_path = os.path.join(os.path.dirname(__file__), "schemas", file_name)
    if not os.path.exists(schema_path):
        raise FileNotFoundError(f"Schema not found: {schema_path}")
    with open(schema_path, "r", encoding="utf-8") as f:
        return json.load(f)


def validate_with_named_schema(data, schema_name):
    schema = load_schema(schema_name)
    return validate_against_schema(data, schema)